### PR TITLE
add test for generic classes

### LINKED ISSUE #75: Generic class model
## Versions

- NodeJS: 10.0.0
- Typegoose(NPM): 6.0.1
- mongoose: 5.7.4
- mongodb: 3.3.2

## Code Example

I'm trying to implement a user model, where there's a different profile subschema based on the role field. The idea was to define role as a discriminator and implement different profile classes for each of them.

```ts
export class DefaultProfile {
  @prop()
  firstName?: string;
}

export class AdminProfile {
  @prop()
  title?: string;
}

@modelOptions({
  schemaOptions: {
    collection: "users",
    discriminatorKey: "role",
    timestamps: true,
  },
})
export class GenericUserSchema<T extends DefaultProfile = DefaultProfile> {
  @prop({ required: true, enum: USER_ROLE })
  role?: USER_ROLE;

  @prop()
  profile?: T;
}

export class UserSchema extends GenericUserSchema<DefaultProfile> {}

export default getModelForClass(UserSchema);

it("should contain additional fields inside the mongoose model", async () => {
    expect(User.schema.obj.profile.type.obj.firstName).toBeTruthy();
});
```

Appearently `getModelForClass(UserSchema)` returns a model which has profile declared as `Mixed` instead of the explicit Model I defined above.

## Do you know *why* it happenes?

Not sure if i'm perhaps misusing ts :disappointed: or if there's a bug in typegoose.

### PR DESCRIPTION
**Pull request by sakulstra**
*targets what was said in #75*

When writing this I wasn't sure any more if this should be supported or not :thinking: as a workaround one could just remove the profile prop from the base model and statically reimplement it on all the descriminators instead of extending a default :thinking: 

Opening it anyways in case someone steps over this.

---

fixes #75

### CODE REVIEW

File: test/tests/discriminators.test.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      role: ROLE.DEFAULT,
+      visitor: 'sth',
+      default: 'sth',
+      profile: { test: 'sth', lastName: 'sth' }
+    } as DisciminatedUser);
+    expect((instance as any).visitor).to.equals(undefined);
+    // why default? this shouldnt exists on the model - at least i cant find it where it would be defined
+    expect(instance.default).to.equals('sth');
+    expect((instance.profile as any).test).to.equals(undefined);
+    // same here, "lastName" is not defined on the "Profile" Model, and the "DefaultProfile" was nowhere defined to be used
```
[the remark this answers is not in the document]
Comment (sakulstra) [reply test/tests/discriminators.test.ts]: same as  above, ROLE.DEFAULT should call the correct discriminator which contains the correct Profile https://github.com/typegoose/typegoose/pull/79/commits/fb96003680b7d417c63f2aad1c523ffa26aafd15#diff-00fa4a2f15549a5406a700563c5a83a3R34

File: test/tests/discriminators.test.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export function suite() {
+  it('should set fields based on role disciminator & ignore other fields', async () => {
+    const instance: DocumentType<DisciminatedUser> = await DisciminatedUserModel.create({
+      role: ROLE.DEFAULT,
+      visitor: 'sth',
+      default: 'sth',
+      profile: { test: 'sth', lastName: 'sth' }
+    } as DisciminatedUser);
+    expect((instance as any).visitor).to.equals(undefined);
+    // why default? this shouldnt exists on the model - at least i cant find it where it would be defined
```
[the remark this answers is not in the document]
Comment (sakulstra) [reply test/tests/discriminators.test.ts]: as `role: ROLE.DEFAULT` is set, the proper discriminator should be called https://github.com/typegoose/typegoose/pull/79/commits/fb96003680b7d417c63f2aad1c523ffa26aafd15#diff-00fa4a2f15549a5406a700563c5a83a3R44

**Comment by hasezoey** [comment]
@sakulstra 
> i will get back to this next week as i'm still not confident that sth is broken.

so, any news on this?

**Comment by hasezoey** [comment]
Identified one issue: the generic "Profile" gets translated by typescript to "Object" which will be interpreted by mongoose as "Mixed" - which in turn allows any property-key value

**Comment by hasezoey** [comment]
does 6884e712c664e4d42c3b49504aa26e1d648a1836 fix this pr? @sakulstra

**Comment by hasezoey** [comment]
now i found the issue again i was searching: #25, is the only thing that is holding this test back

**Comment by sakulstra** [comment]
sorry was on a work trip and completely forgot about this pr :(

Yes you're pr fixes the test indeed :+1:

---

The root cause of the issue I had actually was a thinking mistake(misunderstanding how discriminators work, even without typegoose), not related to this at all :sweat_smile:

**Comment by hasezoey** [comment]
so i will clean this pr, and add merge it (for the extra tests)

**Comment by hasezoey** [comment]
i merged latest master, noticed a "bug" - i will "hijack" this pr for more internal things - because this is the only test i found being reproduceable, hope you dont mind

**Pull request merged**